src/Storages/System/StorageSystemFunctions.cpp

```diff
@@ -143,14 +143,26 @@ void StorageSystemFunctions::fillData(MutableColumns & res_columns, ContextPtr c
         try
         {
             DO_NOT_UPDATE_ERROR_STATISTICS();
-            is_deterministic = functions_factory.tryGet(function_name, context)->isDeterministic();
+            /// some functions try to fetch user in its constructor which can cause some issues when
+            /// user is not set (e.g. when it's executed from DDLWorker)
+            /// because we know return value of `isDeterministic` for those functions we can have
+            /// a lookup table for them
+            static const std::unordered_map<std::string_view, bool> known_values
+            {
+                {"defaultRoles", false},
+                {"defaultProfiles", false},
+            };
+
+            if (auto it = known_values.find(function_name); it != known_values.end())
+                is_deterministic = it->second;
+            else
+                is_deterministic = functions_factory.tryGet(function_name, context)->isDeterministic();
```

Comment:
So the problem seems to be here because `FunctionCurrentRoles` and `FunctionProfiles` both try to get their results in their constructor instead of doing that in their `executeImpl()` functions?

Reply:
Yes actually, we could introduce a flag to initialize on first call.

Reply:
> we could introduce a flag to initialize on first call